src/log_util.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::ffi::CStr;
+
+use grpc_sys::{self, GprLogFuncArgs, GprLogSeverity};
+use log::{self, LogLevel, LogLevelFilter, LogLocation};
+
+#[inline]
+fn severity_to_log_level(severity: GprLogSeverity) -> LogLevel {
+    match severity {
+        GprLogSeverity::Debug => LogLevel::Debug,
+        GprLogSeverity::Info => LogLevel::Info,
+        GprLogSeverity::Error => LogLevel::Error,
+    }
+}
+
+extern "C" fn delegate(c_args: *mut GprLogFuncArgs) {
+    let args = unsafe { &*c_args };
+    let level = severity_to_log_level(args.severity);
+    if !log_enabled!(level) {
+        return;
+    }
+
+    // can't panic.
+    let file_str = unsafe { CStr::from_ptr(args.file).to_str().unwrap() };
+    let line = args.line as u32;
+
+    // use hidden API for now, will
+    // TODO: use public API once available.
+    let location = LogLocation {
```

Comment:
`Loglocation` has been removed (https://github.com/rust-lang-nursery/log/pull/162).

The latest `log` v0.3.8 is release in May 24, and differs  greatly from the master. Any new release of log will break grpcio's build. grpcio and log are tightly coupled if this PR is merged. Crates that use both log and grpcio may fail to build, if log and grpcio are mismatch.

Reply:
Yes. Already consider this. See TODO at L41 and rust-lang-nursery/log#223.